// _rc-tween-one@0.8.5@rc-tween-one/lib/TweenOne.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _util = require('./util');

var _styleUtils = require('style-utils');

var _TimeLine = require('./TimeLine');

var _TimeLine2 = _interopRequireDefault(_TimeLine);

var _plugins = require('./plugins');

var _plugins2 = _interopRequireDefault(_plugins);

var _ticker = require('./ticker');

var _ticker2 = _interopRequireDefault(_ticker);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

function noop() {}

var perFrame = Math.round(1000 / 60);

var tickerIdNum = 0;

var TweenOne = function (_Component) {
  _inherits(TweenOne, _Component);

  function TweenOne() {
    _classCallCheck(this, TweenOne);

    var _this = _possibleConstructorReturn(this, _Component.apply(this, arguments));

    _this.rafID = -1;
    _this.moment = _this.props.moment || 0;
    _this.startMoment = _this.props.moment;
    _this.startFrame = _ticker2["default"].frame;
    _this.paused = _this.props.paused;
    _this.reverse = _this.props.reverse;
    _this.onChange = _this.props.onChange;
    ['raf', 'frame', 'start', 'play', 'restart'].forEach(function (method) {
      return _this[method] = _this[method].bind(_this);
    });
    return _this;
  }

  TweenOne.prototype.componentDidMount = function componentDidMount() {
    this.dom = _reactDom2["default"].findDOMNode(this);
    this.start();
  };

  TweenOne.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    var _this2 = this;

    this.onChange = nextProps.onChange;
    // 跳帧事件 moment;
    var newMoment = nextProps.moment;
    if (typeof newMoment === 'number' && newMoment !== this.moment) {
      this.startMoment = newMoment;
      this.startFrame = _ticker2["default"].frame;
      if (this.rafID === -1 && !nextProps.paused) {
        (function () {
          _this2.timeLine.resetAnimData();
          var style = nextProps.style;
          _this2.dom.setAttribute('style', '');
          Object.keys(style).forEach(function (key) {
            _this2.dom.style[key] = (0, _styleUtils.stylesToCss)(key, style[key]);
          });
          _this2.play();
        })();
      } else {
        this.raf();
      }
    }
    // 动画处理
    var newAnimation = nextProps.animation;
    var currentAnimation = this.props.animation;
    var equal = (0, _util.objectEqual)(currentAnimation, newAnimation);
    var styleEqual = (0, _util.objectEqual)(this.props.style, nextProps.style);
    // 如果 animation 不同， 重新动画
    this.restartAnim = false;
    if (!equal) {
      this.cancelRequestAnimationFrame();
      if (nextProps.resetStyleBool && this.timeLine) {
        this.timeLine.resetDefaultStyle();
      }
      this.startMoment = perFrame - 1; // 设置 perFrame 为开始时就播放一帧动画, 不是从0开始, 鼠标跟随使用
      this.startFrame = _ticker2["default"].frame;
      this.restartAnim = true;
      // this.start(nextProps);
    } else if (!styleEqual) {
      // 如果 animation 相同，，style 不同，从当前时间开放。
      if (this.rafID !== -1) {
        this.cancelRequestAnimationFrame();
        this.startMoment = this.timeLine.progressTime;
        this.startFrame = _ticker2["default"].frame;
        this.restartAnim = true;
        // this.start(nextProps);
      }
    }
    // 暂停倒放
    if (this.paused !== nextProps.paused || this.reverse !== nextProps.reverse) {
      this.paused = nextProps.paused;
      this.reverse = nextProps.reverse;
      if (this.paused) {
        this.cancelRequestAnimationFrame();
      } else {
        if (this.reverse && nextProps.reverseDelay) {
          this.cancelRequestAnimationFrame();
          _ticker2["default"].timeout(this.restart, nextProps.reverseDelay);
        } else {
          this.restart();
        }
      }
    }
  };

  TweenOne.prototype.componentDidUpdate = function componentDidUpdate() {
    // 样式更新了后再执行动画；
    if (this.restartAnim) {
      this.start();
    }
  };

  TweenOne.prototype.componentWillUnmount = function componentWillUnmount() {
    this.cancelRequestAnimationFrame();
  };

  TweenOne.prototype.restart = function restart() {
    this.startMoment = this.timeLine.progressTime;
    this.startFrame = _ticker2["default"].frame;
    this.play();
  };

  TweenOne.prototype.start = function start() {
    var props = this.props;
    if (props.animation && Object.keys(props.animation).length) {
      this.timeLine = new _TimeLine2["default"](this.dom, (0, _util.dataToArray)(props.animation), props.attr);
      // 预先注册 raf, 初始动画数值。
      this.raf();
      // 开始动画
      this.play();
    }
  };

  TweenOne.prototype.play = function play() {
    this.cancelRequestAnimationFrame();
    if (this.paused) {
      return;
    }
    this.rafID = 'tween' + Date.now() + '-' + tickerIdNum;
    _ticker2["default"].wake(this.rafID, this.raf);
    tickerIdNum++;
  };

  TweenOne.prototype.frame = function frame() {
    var moment = (_ticker2["default"].frame - this.startFrame) * perFrame + (this.startMoment || 0);
    if (this.reverse) {
      moment = (this.startMoment || 0) - (_ticker2["default"].frame - this.startFrame) * perFrame;
    }
    moment = moment > this.timeLine.totalTime ? this.timeLine.totalTime : moment;
    moment = moment <= 0 ? 0 : moment;
    if (moment < this.moment && !this.reverse) {
      this.timeLine.resetDefaultStyle();
    }
    this.moment = moment;
    this.timeLine.onChange = this.onChange;
    this.timeLine.frame(moment);
  };

  TweenOne.prototype.raf = function raf() {
    this.frame();
    if (this.moment >= this.timeLine.totalTime && !this.reverse || this.paused || this.reverse && this.moment === 0) {
      return this.cancelRequestAnimationFrame();
    }
  };

  TweenOne.prototype.cancelRequestAnimationFrame = function cancelRequestAnimationFrame() {
    _ticker2["default"].clear(this.rafID);
    this.rafID = -1;
  };

  TweenOne.prototype.render = function render() {
    var props = (0, _objectAssign2["default"])({}, this.props);
    ['animation', 'component', 'reverseDelay', 'attr', 'paused', 'reverse', 'moment', 'resetStyleBool'].forEach(function (key) {
      return delete props[key];
    });
    props.style = (0, _objectAssign2["default"])({}, this.props.style);
    for (var p in props.style) {
      if (p.indexOf('filter') >= 0 || p.indexOf('Filter') >= 0) {
        // ['Webkit', 'Moz', 'Ms', 'ms'].forEach(prefix=> style[`${prefix}Filter`] = style[p]);
        var transformArr = ['Webkit', 'Moz', 'Ms', 'ms'];
        for (var i = 0; i < transformArr.length; i++) {
          props.style[transformArr[i] + 'Filter'] = props.style[p];
        }
      }
    }
    props.component = typeof props.component === 'function' ? this.props.componentReplace : props.component;
    if (!props.component) {
      delete props.component;
    }
    return _react2["default"].createElement(this.props.component, props);
  };

  return TweenOne;
}(_react.Component);

var objectOrArray = _react.PropTypes.oneOfType([_react.PropTypes.object, _react.PropTypes.array]);

TweenOne.propTypes = {
  component: _react.PropTypes.any,
  componentReplace: _react.PropTypes.string,
  animation: objectOrArray,
  children: _react.PropTypes.any,
  style: _react.PropTypes.object,
  paused: _react.PropTypes.bool,
  reverse: _react.PropTypes.bool,
  reverseDelay: _react.PropTypes.number,
  moment: _react.PropTypes.number,
  attr: _react.PropTypes.string,
  onChange: _react.PropTypes.func,
  resetStyleBool: _react.PropTypes.bool
};

TweenOne.defaultProps = {
  component: 'div',
  reverseDelay: 0,
  attr: 'style',
  onChange: noop
};
TweenOne.plugins = _plugins2["default"];
exports["default"] = TweenOne;
module.exports = exports['default'];
