// _slate-react@0.6.0@slate-react/lib/utils/find-selection.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _selectionIsBackward = require('selection-is-backward');

var _selectionIsBackward2 = _interopRequireDefault(_selectionIsBackward);

var _slate = require('slate');

var _findPoint = require('./find-point');

var _findPoint2 = _interopRequireDefault(_findPoint);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Find a Slate range from a DOM `nativeSelection`.
 *
 * @param {Selection} nativeSelection
 * @param {State} state
 * @return {Range}
 */

function findSelection(nativeSelection, state) {
  var anchorNode = nativeSelection.anchorNode,
      anchorOffset = nativeSelection.anchorOffset,
      focusNode = nativeSelection.focusNode,
      focusOffset = nativeSelection.focusOffset,
      isCollapsed = nativeSelection.isCollapsed;

  var anchor = (0, _findPoint2.default)(anchorNode, anchorOffset, state);
  var focus = isCollapsed ? anchor : (0, _findPoint2.default)(focusNode, focusOffset, state);
  if (!anchor || !focus) return null;

  var range = _slate.Range.create({
    anchorKey: anchor.key,
    anchorOffset: anchor.offset,
    focusKey: focus.key,
    focusOffset: focus.offset,
    isBackward: (0, _selectionIsBackward2.default)(nativeSelection),
    isFocused: true
  });

  return range;
}

/**
 * Export.
 *
 * @type {Function}
 */

exports.default = findSelection;
